src/utils/hook.ts

```diff
@@ -250,11 +250,24 @@ export function usePaginationParamsFromSearch(opts: {
     [maxPageSize, updateSearchParams],
   )
 
+  const watchServerPageSize = useCallback(
+    (newPageSize?: number) => {
+      // eslint-disable-next-line react-hooks/rules-of-hooks
```

Comment:
Can the useEffect hook be moved outside of useCallback? This way, it ensures that there won't be any errors or warnings related to React Hooks rules.

Reply:
Putting hooks inside hooks is deliberate. because we need get `setPageSize` reference.

if we move out from the outer hooks. the hook seems like
```ts
useServerPageSize(setPageSize: (val: number) => void, pageNumber) {
  useEffect(() => {
}, []);
}
```
Seems more strange which needed to pass setPageSize to the hook parameter.

I have no idea. Could you please give me a better approach?

Reply:
I would prefer a straight and clear solution, like passing the page size into the pagination, or calculating the total page in `react-query`. It may change a lot, but it will be clear.

Hooks in nested function is not allowed

Ref: https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

Reply:
OK, the whole hooks should be refactored. not only finish this feature requirement.